File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *


input = stdin.readline

class Edge:

    
    def __init__(self,a,b,cost):
        self.a = a
        self.b = b
        self.cost = cost
        

    def __lt__(self, other): # x < y 를 판단하는 기준을 정의 (less than → lt)
        return self.cost < other.cost
    # def __le__(self, other): # x ≤ y 를 판단하는 기준을 정의 (less than or equal to → le)
    #     return self.cost <= other.cost
    # def __gt__(self, other): # x > y 를 판단하는 기준을 정의 (greater than → gt)
    #     return len(self) > len(other)
    # def __ge__(self, other): # x ≥ y 를 판단하는 기준을 정의 (greater than or equal to → ge)
    #     return len(self) >= len(other)
    # def __eq__(self, other): #  x == y 를 판단하는 기준을 정의 (equal to → eq)
    #     return len(self) == len(other)
    # def __ne__(self, other):  x == y 를 판단하는 기준을 정의 (equal to → eq)
    #     return len(self) != len(other)




def union(a,b):
    global root
    pa = find(a)
    pb = find(b)
    
   
    if(pa<pb): # pa가 작으면 
        root[pb]=pa #pb의 부모를 pa로
    else:
        root[pa]=pb #pa의 부모를 pb
        

def find(a) -> int:
    
    global root
    
    #root 찾기
    r=root[a]
    while(r!=root[r]): #부모를 계속 접근
        r=root[r]
    #root 찾음

    #path compression(시간 복잡도를 나중에 갈수록 줄이기 위해)
    c:int=a
    while(c!=root[c]): #root 도달 전까지
        pc=root[c] # 현재 c의 부모를 가르킴
        root[c]=r # c의 부모를 위에서 구한 최상위 root를 가르킴
        c=pc #c를 부모로 옮김

    return r
    
    
    
    
    
    
    

v,e = map(int,input().split())

root = [i for i in range(v+1)] # i노드의 부모노드

heap = []

for _ in range(e):
    a,b,cost = map(int,input().split())
    heappush(heap,Edge(a,b,cost))



ans = 0

while heap:
    
    edge = heappop(heap)
    
    a = edge.a
    b = edge.b
    c = edge.cost
    
    if find(a) != find(b):
        union(a,b)
        ans += c
        
        
        
    
print(ans)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.swift
import Foundation

let ve = readLine()!.split{$0 == " "}.compactMap({Int(String($0))!})

let v = ve[0]
let e = ve[1]

var graph:[(Int,Int,Int)] = [(Int,Int,Int)]()

var parents:[Int] = Array(0...v)

func union(_ a:Int,_ b:Int){
    
    let pa = find(a)
    let pb = find(b)
    
    // 작은게 부모가 된다
    if pa < pb  {
        parents[pb] = pa
    }
    
    else {
        parents[pa] = pb
    }
    
}

func find(_ c:Int) -> Int {
    
    var r = parents[c]
    
    // 보모를 찾음
    while r != parents[r] {
            
        r = parents[r]

    }
    
    var c = c
    
    while c != parents[c] {
        
        let nextC = parents[c]
        parents[c] = r
        c = nextC
    }
    
    
    return r
    
}


for _ in 0..<e {
    
    let info = readLine()!.split{$0 == " "}.compactMap{Int(String($0))!}
    
    graph.append((info[0],info[1],info[2]))

}


graph.sort(by: {
    return $0.2 < $1.2
}) // 가중치로 정렬

var lines:Int = 0
var result:Int = 0

for edge in graph {
    
    if lines == v-1 { // 더한 간선은 항상 정점-1 개이다.
        break
    }
    
    if find(edge.0) != find(edge.1){
        union(edge.0,edge.1)
        result += edge.2
        lines += 1
    }
    
}


print(result)






